mesonbuild/backend/ninjabackend.py

```diff
@@ -2040,17 +2040,20 @@ def generate_rust_target(self, target: build.BuildTarget) -> None:
                 d = '.'
             args += ['-L', d]
         target_deps = target.get_dependencies()
-        has_shared_deps = any(isinstance(dep, build.SharedLibrary) for dep in target_deps)
-        if isinstance(target, build.SharedLibrary) or has_shared_deps:
-            has_rust_shared_deps = any(isinstance(dep, build.SharedLibrary) and dep.uses_rust()
-                                       and dep.rust_crate_type not in {'cdylib', 'proc-macro'}
-                                       for dep in target_deps)
-            if cratetype not in {'cdylib', 'proc-macro'} or has_rust_shared_deps:
-                # add prefer-dynamic if any of the Rust libraries we link
-                # against are dynamic or this is a dynamic library itself,
-                # otherwise we'll end up with multiple implementations of crates
-                args += ['-C', 'prefer-dynamic']
+        has_shared_deps = any(isinstance(dep, build.SharedLibrary)
+                              and dep.rust_crate_type != 'proc-macro'
+                              for dep in target_deps)
+        has_rust_shared_deps = any(dep.uses_rust()
+                                   and dep.rust_crate_type == 'dylib'
+                                   for dep in target_deps)
+
+        if cratetype in {'dylib', 'proc-macro'} or has_rust_shared_deps:
+            # add prefer-dynamic if any of the Rust libraries we link
+            # against are dynamic or this is a dynamic library itself,
+            # otherwise we'll end up with multiple implementations of libstd.
+            args += ['-C', 'prefer-dynamic']
 
+        if isinstance(target, build.SharedLibrary) or has_shared_deps:
```

Comment:
I'm not sure anymore about that condition. For example, a proc-macro is a SharedLibrary but should not need rpath, does it? Another case is a rlib that has shared deps, StaticLibrary cannot have rpath.

What about `cratetype in {'dylib', 'cdylib', 'bin'} and has_shared_deps` ?

Reply:
Why wouldn't a proc macro need rpath?  It's still a shared object, so what makes it different from any other?

Reply:
rpath is used by Meson to run programs from the builddir, to find shared libraries in different subdirs, it replaces setting `LD_LIBRARY_PATH`. proc-macro is not used at runtime, but thinking more about it, rustc will have to load them, so probably that requires rpath at build time.

Maybe just need to replace the `or` to `and` then?

Reply:
> Maybe just need to replace the `or` to `and` then?

ah no, we need to cover executable too...

`if isinstance(target, (build.SharedLibrary, build.Executable)) and has_shared_deps:` ?

Reply:
If the executable doesn't have any shared dependencies, what do we need to set the rpath for?  The only things it dynamically links (at least on my system) are linux-vds, libgcc_s, and libc.

Reply:
> If the executable doesn't have any shared dependencies, what do we need to set the rpath for?

We don't need rpath in that case, which is why I think a `and` is better here.

Maybe I'm overthinking this, I see that for non-rust libraries we just unconditionally call `build_rpath_args()`. `determine_rpath_dirs()` just return empty list if there are no shared deps anyway.

Reply:
Does that mean we could eliminate the conditional entirely, or would we still need to avoid calling `build_rpath_args()` for static libs?

Reply:
It actually makes sense for proc-macro to have rpath pointing to other proc-macros. Maybe I've been misleading you from the beginning, I'm going full circle lol.

Now I think the proper condition would be:

```python
has_shared_deps = any(isinstance(dep, build.SharedLibrary) for dep in target_deps)
if isinstance(target, (build.SharedLibrary, build.Executable)) and has_shared_deps:
```

But that's just a simple optimization to not have to compute an rpath when not needed: when we build a static lib or when it has not shared deps.

Reply:
We definitely still should be excluding proc macros from `has_shared_deps`, because an executable that only depends on a proc macro doesn't need anything in its rpath.

But if it's this difficult to understand, shouldn't we just drop the optimisation and run this code unconditionally?  Is it really worth this level of confusion when AIUI it would continue to do the right thing with no condition here at all?

Reply:
> We definitely still should be excluding proc macros from `has_shared_deps`, because an executable that only depends on a proc macro doesn't need anything in its rpath.

Executables does not need rpath to proc-marco indeed, but proc-macro could need rpath to other proc-macro, no? Makes me wonder how proc-macro works on Windows which does not have rpath... But it opens the door to a another issue, rpath is computed by `backend.rpaths_for_non_system_absolute_shared_libraries()` which does not exclude proc-macro AFAIK...

OTOH, this is a bit offtopic to the original issue...

Reply:
I've just changed this back to what it was before this PR.

Reply:
My understanding is that proc-macros don't depend on each other -- each proc-macro exists in its own little bubble, and is individually loaded into the `rustc` compiler process as a loadable plugin.

Reply:
That's correct